### ISSUE TITLE
properly indicate port state

### ISSUE THREAD

**Issue by rzezeski** [oxidecomputer/opte#92]
With the transition to xde the `opteadm list-ports` command no longer correctly reports if a port is "IN USE" or not. As we have plans to support some type of paused/quiesce state for the purpose of live migration support I think it makes sense to expand this field a bit while fixing it. I think it should report "Active" or "Inactive" to mark the current Port states.

Currently xde creates the Port and immediately places it in the Active state. I think the behavior should be to keep it in the Inactive state until the xde device's `mc_start()` is called, and keep it in the Active state until `mc_stop()` is called (with the exception that once we have a Paused state that will occur while there are active clients). These callbacks indicate when the first client opens the device and when the last client closes the device. Doing it this way allows us to program the Port while its in the Inactive state up until the time the first client starts using it. This gives consumers like Sled Agent more freedom in how it configures an OPTE Port, allowing it to make multiple calls to configure an Inactive port instead of requiring that all bits of configuration be done upfront at xde Port creation time.

**Comment by rzezeski** [comment]
While working on this I realized that the type state pattern re PortState was fighting me more than helping. The state of the port is inherently dynamic, and since we hand a raw pointer off to the mac framework it's not always possible to know what state the references port is in. E.g., the `mc_stop()` callback could pass a port in the running state, or perhaps the paused state (though no real determinations have been made on how pausing actually works yet). Furthermore, the transition from Inactive -> Active was always a one-way street: if you want a port with a different layer pipeline, then make a new one.

My solution was to introduce a `PortState` field that is currently protected by a mutex. A given port operation will now check that the port is in a valid state for that operation, e.g., `Port::process()` will fail unless the state is `Running`. The current life cycle of a port looks like the following:

```
create_xde() -> PortBuilder::create() -> PortState::Ready
mc_start() -> Port::start() -> PortState::Running
mc_stop() -> Port::reset() -> PortState::Ready (this transition clears all flow state, but leaves layers/rules)
delete_xde() -> remove Port and drop
```

I've also added Paused/Restored states, but they currently mean nothing as no code has been written around pause/resume or save/restore. Those ideas are going to require some more consideration about how that's all going to work. But it seemed good to define them as a reminder that they will need to be considered going forward.

**Comment by rzezeski** [comment]
Tested this using a modified otpe-ddm-x2 topo. Transitioned ports between `ready -> running -> ready -> running` while using ping and iperf to send traffic while in running states.

```
root@sled1:/opt/cargo-bay# ./opteadm list-ports                                       
LINK                             MAC ADDRESS              IPv4 ADDRESS     STATE      
xde0                             A8:40:25:FF:00:01        10.0.0.1         running 
xde1                             A8:40:25:FF:00:02        10.0.0.2         ready  
```

Addressed in commit e5bcaa0c6e4b7b39fd9167be235a4ac1153869b4.
